File: 2021/Day12/Day12.scala
import scala.io.Source
import scala.util.chaining._
import scala.language.implicitConversions

@main def day12 =
    case class Route(
        val current: String,
        val visited: Set[String],
        val dupSmall: Option[String]
    )

    def extract(current: String, branch: (String, String)) =
        if branch._1 == current then branch._2
        else branch._1

    def addBranch(node: String, target: String)(map: Map[String, Set[String]]) =
        map.get(node) match
            case None    => map + (node -> Set(target))
            case Some(s) => map + (node -> (s + target))

    def isSmall(node: String) = node.toCharArray.forall(_.isLower)

    def endWith(node: String, branch: (String, String)) = branch._2 == node

    def startWith(node: String, branch: (String, String)) = branch._1 == node

    def goBack(map: Map[String, Set[String]], doubleVisit: Boolean)(
        route: Route
    ): Int =
        val Route(current, visited, dupSmall) = route
        (current, visited.contains(current)) match
            case ("start", _) => 1
            case ("end", _)   => 0
            case (node, true)
                if isSmall(node) && (!doubleVisit || !dupSmall.isEmpty) =>
                0
            case (node, seen) =>
                val newVisited = visited + current
                val newDupSmall = dupSmall match
                    case None =>
                        if seen && doubleVisit && isSmall(node) then Some(node)
                        else None
                    case s @ Some(_) => s

                map.get(current)
                    .get
                    .view
                    .map(next =>
                        goBack(
                          map,
                          doubleVisit
                        )(Route(next, newVisited, newDupSmall))
                    )
                    .sum

    def solve(doubleVisit: Boolean, input: List[(String, String)]) =
        val map = input.foldLeft(Map[String, Set[String]]())((m, branch) =>
            val (l, r) = branch
            m.pipe(addBranch(l, r)).pipe(addBranch(r, l))
        )

        input
            .filter(branch =>
                startWith("end", branch) || endWith("end", branch)
            )
            .map(branch => Route(extract("end", branch), Set("end"), None))
            .map(goBack(map, doubleVisit))
            .sum

    val input = Source
        .fromFile("input.txt")
        .getLines
        .map(s =>
            val Array(l, r) = s.split("-")
            (l, r)
        )
        .toList

    println(s"First part: ${solve(false, input)}")
    println(s"Second part: ${solve(true, input)}")

File: 2021/archived/Day12/Day12.fs
module Day12

open System
open System.IO

type Route =
    { current: string
      visited: string Set
      dupSmall: string option }

let (^^) f g x = f x || g x

let (<->) (s: string) (pattern: string) =
    s.Split([| pattern |], StringSplitOptions.RemoveEmptyEntries)

let extract current (s, e) = if s = current then e else s

let addEdge a b map =
    match Map.tryFind a map with
    | None -> Map.add a (Set.singleton b) map
    | Some (s) -> Map.add a (Set.add b s) map

let isSmall node =
    node |> Seq.toArray |> Array.forall Char.IsLower

let endWith node bracnh = bracnh |> snd |> (=) node

let startWith node bracnh = bracnh |> fst |> (=) node

let rec goBack map doubleVisit route =
    let { current = current
          visited = visited
          dupSmall = dupSmall } =
        route

    match current, Set.contains current visited with
    | "start", _ -> 1
    | "end", _ -> 0
    | node, true when
        isSmall node
        && (not doubleVisit || dupSmall.IsSome)
        ->
        0
    | node, seen ->
        let visited = Set.add current visited

        let dupSmall =
            match dupSmall with
            | None ->
                if seen && doubleVisit && isSmall node then
                    Some(current)
                else
                    None
            | s as Some (_) -> s

        Map.find current map
        |> Seq.sumBy (fun next ->
            goBack
                map
                doubleVisit
                { current = next
                  visited = visited
                  dupSmall = dupSmall })

let solve doubleVisist input =
    let map =
        input
        |> List.fold (fun m (l, r) -> m |> addEdge l r |> addEdge r l) Map.empty

    input
    |> List.filter (startWith "end" ^^ endWith "end")
    |> List.map (fun branch ->
        { current = extract "end" branch
          visited = Set.singleton "end"
          dupSmall = None })
    |> List.map (goBack map doubleVisist)
    |> List.sum

let input =
    File.ReadAllLines "input.txt"
    |> Array.map (fun s ->
        let pair = s <-> "-"
        (pair.[0], pair.[1]))
    |> Array.toList

solve false input |> printfn "First part: %d"
solve true input |> printfn "Seond part: %d"
